D2 Cluster Subsetting Implementation

rickzx opened the pull request:
Low QPS is a long-standing problem with the D2 client-side non-cooperative load balancing library. When QPS is low to a server host, its load balancing signals are weak, e.g., its average latency might be volatile when comparing with the average cluster latency, resulting in ineffective degradations and recoveries. 

D2 Subsetting approaches this problem by limiting a single client to only access a subset of the downstream hosts, resulting in higher QPS and stronger signals of a downstream host with the same amount of requests from the client.

The key of D2 Subsetting is to maintain an overall balanced traffic distribution to the downstream, even though the clients are picking different subsets from the downstream hosts.

rachelhanhan reviewed (commented):
First pass, will do more detailed review on Friday

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/ZKDeterministicSubsettingMetadataProvider.java:
```diff
@@ -0,0 +1,91 @@
+        List<URI> sortedUris = uriProperties.getPartitionDesc().keySet().stream()
+            .filter(uri -> uri.getScheme().equals(_nodeUri.getScheme()))
+            .sorted()
+            .collect(Collectors.toList());
```
Good suggestion! I changed the implementation so that it uses the clusterGenerationId to check for staleness metadata. Now it only goes through the sorting logic when the peer cluster is updated, i.e. version number changes.

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/DeterministicSubsettingStrategy.java:
```diff
@@ -0,0 +1,315 @@
+    for (Map.Entry<T, Double> entry: pointsMap.entrySet())
+    {
+      if (!_currentPointsMap.containsKey(entry.getKey()) ||
+          !isEqual(_currentPointsMap.get(entry.getKey()), entry.getValue(), POINTS_MAP_CHANGED_DELTA))
```
Agreed. I now uses the clusterGenerationId to check for staleness. It should cover the 2 cases you listed above. The clusterGenerationId is updated when URI/Cluster/Service/PartitionProperties are changed. And since cluster/service/partition properties are changed rarely, most of the times, we will regenerate subset map when URIs are updated.

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/SubsettingStrategy.java:
```diff
@@ -0,0 +1,31 @@
+{
+  public static final int DEFAULT_CLUSTER_SUBSET_SIZE = -1;
+
+  Map<T, Double> getWeightedSubset(Map<T, Double> pointsMap);
```
Done. I also changed the naming from pointsMap to weightMap to help distinguish the scale. By convention, weight is 0.0-1.0 and point is 0-100.

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerState.java:
```diff
@@ -627,6 +660,43 @@ public void setDelayedExecution(long delayedExecution)
+        {
+          URI uri = entry.getKey();
+          TrackerClient client = potentialClients.get(uri);
+          client.setSubsetWeight(partitionId, subsetMap.get(uri));
```
Fixed. I store a subsetMapCache in the loadBalancerState. When a client is querying the subset, it will compare the serviceName, partitionId, and clusterGenerationId against the cache. So we can avoid setting subset weight again and again for each request.

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/subsetting/DeterministicSubsettingStrategyTest.java:
```diff
@@ -0,0 +1,204 @@
+
+  private static double[] generateRandomWeights(int size)
+  {
+    return new Random().doubles(size, 0D, 1D).toArray();
```
I haven't looked into the detailed calculation implementation yet, just want to double check that we're not making assumption that the weight is always in the scale of 0-1. 
I think Dyno might increate the weight to something greater than 1.

rickzx replied:
Yes, it will work with weight in all scales.

bbarkley replied:
Probably worth adding some test cases to verify this

rachelhanhan commented on d2-schemas/src/main/pegasus/com/linkedin/d2/D2Service.pdl:
```diff
@@ -96,4 +96,9 @@ record D2Service includes D2ChangeTimeStamps {
+  /**
+   * The minimum cluster subset size for this service. With a value greater than 0, each client will only send requests to a subset of the hosts in the cluster. A value of -1 disables the feature.
+   */
+  minClusterSubsetSize: optional int
```
Another question about this subsetSize - is this size the number of physical host size or weight size? For example, if a host has the weight of 20.0, in the same cluster there are 10 other hosts with weight 1.0, which means that redlined host should get 2/3 of the traffic. In this way, if the subset size is 5 for example, what does 5 mean here?

rickzx replied:
Strictly speaking, minClusterSubsetSize of _s_ is equivalent to a fixed aperture width that contains _s_ hosts **when all hosts have weight 1.0**. But since this definition contains too many terminology, and d2 weight is 1.0 most of the time, I simplify the description a little bit.

For your example, the minimum subset will occupy (5/30 = 1/6) of the ring. When the host weights change, there may be more/less than 5 hosts in the subset.

rachelhanhan replied:
That makes sense. Thank you

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/ZKDeterministicSubsettingMetadataProvider.java:
```diff
@@ -0,0 +1,113 @@
+  private long _clusterGenerationId = -1;
+  private DeterministicSubsettingMetadata _subsettingMetadata;
+
+  public ZKDeterministicSubsettingMetadataProvider(String clusterName,
+                                      URI nodeUri,
+                                      LoadBalancerState loadBalancerState,
+                                      long timeout,
+                                      TimeUnit unit)
```
I see the problem. There's a circular dependency between ZKDeterministicSubsettingMetadataProvider and LoadBalancerState. I resolved this circular dependency by removing LoadBalancerState from the constructor of ZKDeterministicSubsettingMetadataProvider, and passing it as an argument to ZKDeterministicSubsettingMetadataProvider.getSubsettingMetadata(state). Then have LoadBalancerState depend on SubsettingMetadataProvider directly.

Then in container, with clusterName, nodeUri, timeout and unit, we're able to instantiate ZKDeterministicSubsettingMetadataProvider and pass it to d2 config. D2 config will use the ZKDeterministicSubsettingMetadataProvider to instantiate LoadBalancerState. Then in LoadBalancerState, it will create an instance of SubsettingStrategyFactory using ZKDeterministicSubsettingMetadataProvider and itself.

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/DeterministicSubsettingMetadataProvider.java:
```diff
@@ -0,0 +1,25 @@
+ */
+public interface DeterministicSubsettingMetadataProvider
+{
+  DeterministicSubsettingMetadata getSubsettingMetadata();
```
done

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/ZKDeterministicSubsettingMetadataProvider.java:
```diff
@@ -0,0 +1,113 @@
+    {
+      return metadataFutureCallback.get(_timeout, _unit);
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      return null;
```
Logging is in DeterministicSubsettingStrategy.getWeightedSubset. When the metadata is null, it logs warn("Cannot retrieve metadata required for D2 subsetting. Revert to use all available hosts.");

bbarkley replied:
Yes, but you can't differentiate between a timeout and the result of the callback there. I would think we'd want to know if an exception is being handled here.

rickzx replied:
Added a log event here.

rachelhanhan commented on d2-schemas/src/main/pegasus/com/linkedin/d2/D2Service.pdl:
```diff
@@ -96,4 +96,9 @@ record D2Service includes D2ChangeTimeStamps {
+  /**
+   * The minimum cluster subset size for this service. With a value greater than 0, each client will only send requests to a subset of the hosts in the cluster. A value of -1 disables the feature.
+   */
+  minClusterSubsetSize: optional int
```
I see, I'm not against setting it to d2Service level either, I think both way works.
In the long-term, do you imagine that subsetting being enabled for every service or only for services that suffer from the low QPS problem?

rickzx replied:
Potentially we could enable subsetting for all services, as it has other benefits than solving low QPS, for example, making node failure discovery faster. But we'll need to see via initial piloting whether load distribution or subsetting overhead is as desired.

rickzx replied:
The default values for d2 config are defined in global-config. If the user doesn't explicitly define a config in a service, the default value will be used. So this shouldn't be a problem here.

I separate the control for enabling d2 subsetting and set the subset size. I feel like it is more manageable in this way.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/DeterministicSubsettingStrategy.java:
```diff
@@ -0,0 +1,306 @@
+    return Math.abs(a - b) <= delta;
+  }
+
+  private double getSubsetSliceWidth(int totalClientCount, int totalHostCount)
```
To get the subset slice width, should we use server host total weight instead of total count?

rickzx replied:
It should be total count. As mentioned above, the subset slice width is calculated given all hosts have equal weights. It should not change when the host weights are changed. The subset slice width is the same for situation A {host1 = 1.0, host2 = 1.0, host3 = 1.0} and situation B {host1 = 1.0, host2 = 0.0, host3 = 0.0}.

rachelhanhan replied:
Can you use this example to help me understand how is it calculated:
Server side host: 1 redlined host weight: 10.0, 10 other regular host weights: 1.0, so we have total weights as 20.0
MinSubsetSize: 10
Client side: 4 hosts

In this case, what's the subset slice width? I though it should be 1/2? We can have 4 hosts each cover 1/2 of the ring?

rickzx replied:
In your case it would be the entire ring, i.e. no subsetting effect.

Two things to keep in mind: 1. the slice width is calculated given all hosts have weights 1.0.    2. The slice width is a multiple of client unit width = (1 / client count total)

So in your case, minSubsetSize of 10 corresponds to a slice width of 10/11, since there are 11 hosts in total. But since 10/11 is not a multiple of client unit width = 1/4, we will take the next multiple, which is 1.0.

If we have each client covers only 1/2 of the ring, it does not satisfy the minSubsetSize requirement 10/11 when all the hosts return to normal.

Here's a cool visualization of the algorithm https://nvartolomei.com/weighted-deterministic-aperture/. You can play around with weights by clicking the Verbose State drop menu.

rachelhanhan replied:
What confuses me is that what does minSubsettingSize mean, does it mean the actual host number or the weight total (the same question I asked in the pdl change).
If it always means the actual host number, then this is right.

rickzx replied:
It means the actual host number, but in the form of slice width. For a cluster with 100 hosts, minSubsetSize of 10 means 1/10 slice width. It doesn't change when host weight changes.

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/subsetting/DeterministicSubsettingStrategyTest.java:
```diff
@@ -0,0 +1,208 @@
+  }
+
+  @Test(dataProvider = "differentWeightsData")
+  public void testDistributionWithDifferentWeights(int clientNum, double[] weights, int minSubsetSize)
```
Should we also verify minSubsetSize requirement is met?

rickzx replied:
Done. Added minSubsetSize verifier to the testDistributionWithDifferentWeights test

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/ZKDeterministicSubsettingMetadataProvider.java:
```diff
@@ -0,0 +1,114 @@
+ */
+public class ZKDeterministicSubsettingMetadataProvider implements DeterministicSubsettingMetadataProvider
+{
+  private final Logger _log = LoggerFactory.getLogger(ZKDeterministicSubsettingMetadataProvider.class);
```
Not used

rickzx replied:
Removed

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/SubsettingStrategyFactory.java:
```diff
@@ -0,0 +1,32 @@
+   * get retrieves the {@link SubsettingStrategy} corresponding to the serviceName and partition Id
+   * @return {@link SubsettingStrategy}
+   */
+   SubsettingStrategy<URI> get(String serviceName, ServiceProperties serviceProperties, int partitionId);
```
Fixed.

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerState.java:
```diff
@@ -627,6 +671,55 @@ public void setDelayedExecution(long delayedExecution)
+    if (subsettingStrategy != null)
+    {
+      if (subsettingStrategy.isSubsetChanged(_version.get()) ||
+          !_weightedSubsetsCache.containsKey(serviceName) ||
+          !_weightedSubsetsCache.get(serviceName).containsKey(partitionId))
```
I would prefer to decouple subsetting strategy from the simpleLoadBalancerState caching logic. First, I would like to keep subsetting strategy as a general interface, as to pick a subset from a collection of items. This is also how the D2 Hash Ring works, as the interface suggests 

public interface Ring<T>
{
  /**
   * Pick an object in the ring given a specified key
   *
   * @param key   The value used to pick an object from the hash ring
   * @return An object in the ring
   */
  @Nullable
  T get(int key);
}

Second, the caching logic also requires setting subset weight on TrackerClient, which I think is specific to SimpleLoadBalancerState.

As for exposing isSubsetChanged as a public interface, we need this because we don't want to set TrackerClient subset weight over and over again when the weight map doesn't change. If the subsetting strategy only returns the subset, the SimpleLoadBalancer has no idea whether the subset is changed from previous one.

rachelhanhan replied:
Alright, I'm convinced, when there're multiple implementations of SubsettingStrategy, the cache management logic should be the same.

rachelhanhan reviewed (commented):
Looks good to me! Please also update the version once it's finalized.

zizhong commented on d2-schemas/src/main/pegasus/com/linkedin/d2/D2Service.pdl:
```diff
@@ -96,4 +96,9 @@ record D2Service includes D2ChangeTimeStamps {
   backupRequests: optional array[BackupRequestsConfiguration]
+
+  /**
+   * The minimum cluster subset size for this service. With a value greater than 0, each client will only send requests to a subset of the hosts in the cluster. A value of -1 disables the feature.
```
The document seems missing the case of the value being 0.

rickzx replied:
A value of 0 will be considered as invalid, and will be caught by config validations in d2-c-s.

bbarkley reviewed (commented):
Off to a good start here. I still need to go through some of the strategy details if I have time.

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/clients/TrackerClientImpl.java:
```diff
@@ -73,6 +74,7 @@
   private final URI _uri;
   private final Predicate<Integer> _isErrorStatus;
   private final boolean _doNotSlowStart;
+  private final Map<Integer, Double> _subsetWeightMap;
```
Fixed

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/properties/ServiceProperties.java:
```diff
@@ -254,8 +281,8 @@ public String toString()
         + _backupRequests
-        + ", relativeStrategyProperties="
-        + _relativeStrategyProperties
+        + ", minimumClusterSubsetSize="
+        + _minClusterSubsetSize
```
No, in the original code, there's a duplication in logging relativeStrategyProperties. It is already presented in the previous lines. I'm fixing it here

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerState.java:
```diff
@@ -138,6 +143,9 @@
   private final SslSessionValidatorFactory _sslSessionValidatorFactory;
 
+  private final SubsettingStrategyFactory _subsettingStrategyFactory;
+  private final Map<String, Map<Integer, Map<URI, TrackerClient>>>                      _weightedSubsetsCache;
```
In this SimpleLoadBalancerState class, there are a couple other existing nested ConcurrentMaps with similar declared types, e.g. `private final Map<String, LoadBalancerStrategyFactory<? extends LoadBalancerStrategy>> _loadBalancerStrategyFactories;`

I think I would just follow the existing conventions and declare it as `Map<String, Map<Integer, Map<URI, TrackerClient>>>` in this class.

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerState.java:
```diff
@@ -627,6 +671,55 @@ public void setDelayedExecution(long delayedExecution)
+                                                  Map<URI, TrackerClient> potentialClients)
+  {
+    SubsettingStrategy<URI> subsettingStrategy = _subsettingStrategyFactory.get(serviceName, minClusterSubsetSize, partitionId);
+    if (subsettingStrategy != null)
```
Fixed

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/SubsettingStrategyFactoryImpl.java:
```diff
@@ -0,0 +1,71 @@
+  {
+    if (minClusterSubsetSize > 0)
+    {
+      if (_subsettingStrategyMap.containsKey(serviceName))
```
Fixed

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/SubsettingStrategyFactoryImpl.java:
```diff
@@ -0,0 +1,71 @@
+      if (_subsettingStrategyMap.containsKey(serviceName))
+      {
+        Map<Integer, SubsettingStrategy<URI>> strategyMap = _subsettingStrategyMap.get(serviceName);
+        if (strategyMap.containsKey(partitionId))
```
Fixed

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/ZKDeterministicSubsettingMetadataProvider.java:
```diff
@@ -0,0 +1,112 @@
+
+  @GuardedBy("_lock")
+  private long _clusterGenerationId = -1;
+  private DeterministicSubsettingMetadata _subsettingMetadata;
```
Fixed

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerState.java:
```diff
@@ -627,6 +671,55 @@ public void setDelayedExecution(long delayedExecution)
+    {
+      if (subsettingStrategy.isSubsetChanged(_version.get()) ||
+          !_weightedSubsetsCache.containsKey(serviceName) ||
+          !_weightedSubsetsCache.get(serviceName).containsKey(partitionId))
```
Fixed

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerState.java:
```diff
@@ -627,6 +671,55 @@ public void setDelayedExecution(long delayedExecution)
+        {
+          return potentialClients;
+        }
+        else
```
Fixed

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerState.java:
```diff
@@ -627,6 +671,55 @@ public void setDelayedExecution(long delayedExecution)
+            subsetClients.put(uri, client);
+          }
+
+          _weightedSubsetsCache.putIfAbsent(serviceName, new ConcurrentHashMap<>());
```
Fixed using computeIfAbsent

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/DeterministicSubsettingStrategy.java:
```diff
@@ -0,0 +1,306 @@
+import java.util.Map;
+import java.util.Random;
+import java.util.stream.Collectors;
+import org.apache.http.annotation.GuardedBy;
```
Fixed

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/DeterministicSubsettingStrategy.java:
```diff
@@ -0,0 +1,306 @@
+  @GuardedBy("_lock")
+  private DeterministicSubsettingMetadata _currentMetadata;
+  private long _currentVersion = -1;
+  private Map<T, Double> _currentWeightedSubset;
```
Fixed

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/SubsettingStrategyFactoryImpl.java:
```diff
@@ -0,0 +1,71 @@
+
+public class SubsettingStrategyFactoryImpl implements SubsettingStrategyFactory
+{
+  private final Map<String, Map<Integer, SubsettingStrategy<URI>>> _subsettingStrategyMap;
```
Fixed

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/SubsettingStrategyFactoryImpl.java:
```diff
@@ -0,0 +1,71 @@
+  @Override
+  public SubsettingStrategy<URI> get(String serviceName, int minClusterSubsetSize, int partitionId)
+  {
+    if (minClusterSubsetSize > 0)
```
Fixed

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/zkfs/ZKFSTogglingLoadBalancerFactoryImpl.java:
```diff
@@ -34,6 +34,9 @@
 import com.linkedin.d2.balancer.strategies.LoadBalancerStrategyFactory;
+import com.linkedin.d2.balancer.subsetting.DeterministicSubsettingMetadataProvider;
+import com.linkedin.d2.balancer.subsetting.SubsettingStrategyFactory;
+import com.linkedin.d2.balancer.subsetting.SubsettingStrategyFactoryImpl;
```
Fixed

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/D2ClientConfig.java:
```diff
@@ -21,6 +21,9 @@
 import com.linkedin.d2.balancer.strategies.LoadBalancerStrategyFactory;
+import com.linkedin.d2.balancer.subsetting.DeterministicSubsettingMetadataProvider;
+import com.linkedin.d2.balancer.subsetting.SubsettingStrategyFactory;
+import com.linkedin.d2.balancer.subsetting.SubsettingStrategyFactoryImpl;
```
Fixed

rickzx commented on d2/src/main/java/com/linkedin/d2/balancer/D2ClientBuilder.java:
```diff
@@ -33,6 +33,8 @@
 import com.linkedin.d2.balancer.strategies.relative.RelativeLoadBalancerStrategy;
 import com.linkedin.d2.balancer.strategies.relative.RelativeLoadBalancerStrategyFactory;
+import com.linkedin.d2.balancer.subsetting.DeterministicSubsettingMetadataProvider;
+import com.linkedin.d2.balancer.subsetting.SubsettingStrategyFactory;
```
Fixed

rickzx commented:
Thanks for the code review @rachelhanhan and @bbarkley. I addressed all the issues you raised here. Please help take another look when you got time. Thank you!

bbarkley commented on d2/src/main/java/com/linkedin/d2/balancer/subsetting/SubsettingStrategyFactory.java:
```diff
@@ -0,0 +1,32 @@
+  SubsettingStrategyFactory NO_OP_SUBSETTING_STRATEGY_FACTORY = (serviceName, minClusterSubsetSize, partitionId) -> null;
+
+  /**
+   * get retrieves the {@link SubsettingStrategy} corresponding to the serviceName and partition Id
```
Should clarify under which conditions this will return null

bbarkley reviewed (approved):
Sorry for the slow review process on this. Looks good - would recommend clarifying the Javadoc as noted below

rachelhanhan reviewed (approved):
Remember to bump minor version
